TACC-Cloud/geoapi
# task/WG-236: Websocket notifications

## Overview: ##
Backend changes for websocket notifications. 
## Related Jira tickets: ##

* [WG-236](https://tacc-main.atlassian.net/browse/WG-236)

## Summary of Changes: ##

- Rather than using constant polling for notifications, I made changes to use a websocket solution. This PR/branch is to be used with the react front-end's branch that's named the same.
- Added socket io package with poetry
- In the app.py file added different events for socket to listen for, and respond appropriately.
- Added CORS allowed origin for localhost since client and socket server are on different ports.
- Changed number of workers for gunicorn from 4 to 1 bc documentation says that gunicorn only supports 1 worker when using websockets
- Changed nginx.conf (local version) and its rules to handle socket.io path. This allowed for bi-directional handshake between client and server, and websocket connection to remain stable and continuously connected. 

## Testing Steps: ##
1. See hazmapper PR branch with identical name for testing steps. [Front-end](https://github.com/TACC-Cloud/hazmapper/pull/228)

## Notes: ##
- This feature still needs wss security added before being deployed to prod. This can be done in nginx.conf file by adding the SSL cert and key that's related to prod VM (Nathan did this in PR WG-270, so most of the work that's left is on the front-end. However you'll need to set up nginx.conf to upgrade the http connection to handle websockets, I did this in the local version of nginx on this PR for reference). Then on the front-end change socket.io request (socketUtils file) from 'http://localhost:8888' to 'wss://hazmapper.tacc.utexas.edu' (or wherever nginx is running for prod env--might be prod.geoapi-services.tacc.utexas.edu? Note since we moved the port to 443 instead of 8888, we won't need to include the port bc wss defaults to 443).
- Auth can also be changed, if desired. Right now the user must be logged in to get a token and thus establish a web socket connection with server. However, it's using the Bearer Token for auth/verification. If devs want to use JWT instead, you can use extraHeaders option in initial http request to set the JWT [example code using extraHeaders](https://github.com/socketio/engine.io-client#nodejs-with-extraheaders)
